emmarin/cl/cl/src/cl/sparse_merkle.rs

```diff
@@ -0,0 +1,360 @@
+use std::collections::BTreeSet;
+
+use crate::cl::merkle;
+use lazy_static::lazy_static;
+
+/// absence of element is marked with all 0's
+pub static ABSENT: [u8; 32] = [0u8; 32];
+
+/// presence of element is marked with all 1's
+pub static PRESENT: [u8; 32] = [255u8; 32];
+
+lazy_static! {
+    // the roots of empty merkle trees of diffent heights
+    // i.e. all leafs are ABSENT
+    static ref EMPTY_ROOTS: [[u8; 32]; 257] = {
+        let mut roots = [ABSENT; 257];
+        for h in 1..257 {
+            roots[h] = merkle::node(roots[h - 1], roots[h - 1]);
+        }
+
+        roots
+    };
+}
+
+pub fn sparse_root(elems: &BTreeSet<[u8; 32]>) -> [u8; 32] {
+    sparse_root_rec(0, elems)
+}
+
+fn sparse_root_rec(prefix: u64, elems: &BTreeSet<[u8; 32]>) -> [u8; 32] {
+    if elems.is_empty() {
+        return empty_tree_root(256 - prefix);
+    }
+    if prefix == 256 {
+        assert_eq!(elems.len(), 1);
+        return PRESENT;
+    }
+    // partition the elements
+    let (left, right): (BTreeSet<_>, BTreeSet<_>) =
+        elems.iter().partition(|e| !bit(prefix as u8, **e));
```

Comment:
I guess if we use `Box`s or other pointers we could avoid cloning all the elements, althought this is not on the hot path in zk

Reply:
Yeah probably, I did not optimize this beyond pre-computing the empty trees. It's fast enough for now. Also not worth optimizing in case we switch to IMT's